### PR TITLE
[7576]: Support for `vscode.workspace.findTextInFiles` API

### PR DESCRIPTION
**Pull request by nmorenor**
#### What it does

- fix https://github.com/eclipse-theia/theia/issues/7576: Provides ability to execute the vscode api vscode.workspace.findTextInFiles

#### How to test

This was tested with the Codetogether extension since it does use this api.

1. Install the extension
2. Start a session via command palette.
3. On the remote session on browser go to the search view.
4. Search for some text in the workspace you should see search text results.

Alternatively you can use a test extension (https://github.com/Genuitec/theiatestsearch):
1. Install the extension
2. Invoke the SearchTest: Search from command palette and let the input box know what you are searchig for.
3. On the Search test tree view contribution you should see the search results

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by nmorenor** [comment]
This fixes https://github.com/eclipse-theia/theia/issues/7576

File: packages/plugin/src/theia.d.ts [hunk 1]
```diff
@@ -4758,6 +4758,16 @@ declare module '@theia/plugin' {
          */
         export function findFiles(include: GlobPattern, exclude?: GlobPattern | null, maxResults?: number, token?: CancellationToken): PromiseLike<Uri[]>;
 
+        /**
+         * Find text inf files across all [workspace folders] in the workspace
```
Comment (vince-fugnitto) [inline packages/plugin/src/theia.d.ts:4762]: typo: `inf` > `in`

File: packages/plugin/src/theia.d.ts [hunk 2]
```diff
@@ -5006,6 +5016,133 @@ declare module '@theia/plugin' {
         constructor(base: WorkspaceFolder | string, pattern: string)
     }
 
+    /**
```
Comment (vince-fugnitto) [inline packages/plugin/src/theia.d.ts:5019]: I believe these changes should exist in `theia-proposed.d.ts` and not `theia.d.ts` as this API is still proposed in vscode and not finalized.

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts [hunk 3]
```diff
@@ -43,12 +49,16 @@ export class SearchInWorkspaceClientImpl implements SearchInWorkspaceClient {
 
 export type SearchInWorkspaceCallbacks = SearchInWorkspaceClient;
 
+/**
+ * Service to search text in the workspace files.
+ */
+
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-service.ts:55]: Leftover?

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts [hunk 4]
```diff
@@ -43,12 +49,16 @@ export class SearchInWorkspaceClientImpl implements SearchInWorkspaceClient {
...
 /**
  * Service to search text in the workspace files.
  */
 
 @injectable()
-export class SearchInWorkspaceService implements SearchInWorkspaceClient {
+export class SearchInWorkspaceService implements SearchInWorkspaceClient, ISearchInWorkspaceService {
```
Comment (akosyakov) [inline packages/search-in-workspace/src/browser/search-in-workspace-service.ts:61]: boilerplate it not encouraged: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#interfacessymbols

please inject `SearchInWorkspaceService` directly and add new methods here

File: packages/search-in-workspace/src/common/search-in-workspace-interface.ts [hunk 5]
```diff
@@ -143,3 +144,11 @@ export interface SearchInWorkspaceServer extends JsonRpcServer<SearchInWorkspace
 
     dispose(): void;
 }
+
+export const SearchInWorkspaceServiceExtension = Symbol('SearchInWorkspaceServiceExtension');
+
+export interface ISearchInWorkspaceService {
```
Comment (akosyakov) [inline packages/search-in-workspace/src/common/search-in-workspace-interface.ts:150]: this file to define a JSON-RPC protocol between client and server, it seems to be an extension to frontend service, not a part of protocol.

File: packages/plugin/src/theia.d.ts (hunk 2 above)
Comment (nmorenor) [reply packages/plugin/src/theia.d.ts:5019]: The interfaces used by the search api have been moved to theia-proposed.d.ts

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts (hunk 4 above)
Comment (nmorenor) [reply packages/search-in-workspace/src/browser/search-in-workspace-service.ts:61]: Thanks, I've removed it and now directly injecting the search service.

**Comment by nmorenor** [comment]
I'm not sure what is a CQ but here are the orignal interfaces for the API from VSCode. https://github.com/microsoft/vscode/blob/f5d10326f5ef42eb7b6e3a3ce7a50baa3c7d7ed4/src/vs/vscode.proposed.d.ts these had to be the same so I bring those interfaces to this change.

File: packages/search-in-workspace/src/common/search-in-workspace-interface.ts [hunk 6]
```diff
@@ -127,6 +127,7 @@ export interface SearchInWorkspaceClient {
      */
     onDone(searchId: number, error?: string): void;
 }
+export type SearchInWorkspaceCallbacks = SearchInWorkspaceClient;
```
Comment (akosyakov) [inline packages/search-in-workspace/src/common/search-in-workspace-interface.ts:130]: please revert this file

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts (hunk 4 above)
Comment (akosyakov) [inline packages/search-in-workspace/src/browser/search-in-workspace-service.ts:61]: please remove `ISearchInWorkspaceService` and use `SearchInWorkspaceService` directly, i.e.  `container.get(SearchInWorkspaceService)`

Comment (nmorenor) [reply packages/search-in-workspace/src/browser/search-in-workspace-service.ts]: Weird somehow I managed to push to the wrong branch name, sorry about that.

File: packages/search-in-workspace/src/common/search-in-workspace-interface.ts (hunk 6 above)
Comment (nmorenor) [reply packages/search-in-workspace/src/common/search-in-workspace-interface.ts]: Same here, wrong push from my side

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 7]
```diff
@@ -227,6 +233,59 @@ export class WorkspaceMainImpl implements WorkspaceMain, Disposable {
         return uriStrs.map(uriStr => Uri.parse(uriStr));
     }
 
+    async $findTextInFiles(query: theia.TextSearchQuery, options: theia.FindTextInFilesOptions,
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/workspace-main.ts:236]: JSON RPC call cannot accept functions like `callback`. I am not sure how it works. Could you elaborate please?

That's how remote API looks in VS Code: https://github.com/microsoft/vscode/blob/45aa0bf525b08d7e23c9b10ca65f1de55949e601/src/vs/workbench/api/common/extHost.protocol.ts#L722

**Comment by akosyakov** [comment]
@vince-fugnitto Could you try please?

File: packages/plugin-ext/src/main/browser/workspace-main.ts (hunk 7 above)
Comment (nmorenor) [reply packages/plugin-ext/src/main/browser/workspace-main.ts:236]: Right, this _callback does not have a value it is `undefined` I guess we can remove this parameter. I did not do that, because I saw on the VSCode impl that they do allow to send a function like this. The workaround I added was the usage of this.proxy.$onTextSearchResult instead of using this _callback.

Once we are on $onTextSearchResult it process the result and emit an event `searchInWorkspaceEmitter` that is being listened to run the original `callback`.

```
        const disposable = this.searchInWorkspaceEmitter.event(result => {
            callback(result);
        });
        if (token) {
            token.onCancellationRequested(() => {
                disposable.dispose();
            });
        }
```

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Could you try please?

I tried using the second use-case (the `.vsxi` provided by the following [extension](https://github.com/Genuitec/theiatestsearch)) and it looked to work correctly. I was able to see results in the `search-test widget` when searching for terms.

File: packages/plugin-ext/src/main/browser/workspace-main.ts (hunk 7 above)
Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/workspace-main.ts:236]: ok, please remove unused arguments

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/workspace-main.ts:236]: How does it work with 2 parallel searches? Results reported for both calls, should not we filter them? It would be good to have a way to test it.

Comment (nmorenor) [reply packages/plugin-ext/src/main/browser/workspace-main.ts]: Ok, just remove the _callback argument and you are right we should filter parallel searches, just added a way to do that. I guess that a way to test would just run the test case two at the same time with different queries. The approach is to have a sequence number for the search api request and filter the results based on that.

**Comment by kittaakos** [comment]
Is this the reference? https://github.com/microsoft/vscode/blob/54d4edde505fb23e0094e56ba15a902764384690/src/vs/workbench/api/common/extHostWorkspace.ts#L447-L536

**Comment by nmorenor** [comment]
Right, that will be the implementation on VSCode

**Comment by kittaakos** [comment]
I am verifying it with https://github.com/Genuitec/theiatestsearch.

 - I installed the extension to both Theia and VS Code. I have both Theia and VS Code opened on your PR as a workspace.
 - When I execute the `SearchTest: Search` and search for _index_, I have 152 results with Theia. In VS Code, I see 1242 results.

**Comment by kittaakos** [comment]
> * I have 152 results with Theia.

I see. You are [limiting the hit counts to 150](https://github.com/eclipse-theia/theia/pull/7868/files#diff-5fc7724ca5735ff4b2668a5b66980f31R238). It's a bug then.

File: packages/plugin-ext/src/main/browser/workspace-main.ts [hunk 8]
```diff
@@ -227,6 +233,59 @@ export class WorkspaceMainImpl implements WorkspaceMain, Disposable {
...
+                    const hasSearch = this.workspaceSearch.has(searchId);
+                    if (hasSearch) {
+                        this.workspaceSearch.delete(searchId);
+                    }
+                    if (options.maxResults && options.maxResults < matches) {
+                        resolve({ limitHit: true });
+                    } else if (defaultMaxHits < matches) {
+                        resolve({ limitHit: true });
+                    }
+                    resolve({ limitHit: false });
```
Comment (kittaakos) [inline packages/plugin-ext/src/main/browser/workspace-main.ts:274]: Let's clean this up, and move the `resolve({ limitHit: false });` into an `else`.

**Comment by kittaakos** [comment]
I found another regression, search for [_körte_](https://translate.google.com/#view=home&op=translate&sl=auto&tl=en&text=k%C3%B6rte) with the search-in-workspace (SIW) (<kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd>) you have to see one single hit if you opened Theia as a workspace. Use your test extension in VS Code, and search for _körte_ again, and again. You will have one single result. 👍 Now, do the same in Theia:
screencast 2020-06-26 11-16-06
SIW works in Theia:
screencast 2020-06-26 11-18-29

**Comment by kittaakos** [comment]
Hmm, the hit-count feels random. Search for "foo" with SIW, I see 542 results, do the same with your test extension, I see 160 results.

File: packages/plugin-ext/src/plugin/workspace.ts [hunk 9]
```diff
@@ -106,6 +117,33 @@ export class WorkspaceExtImpl implements WorkspaceExt {
...
+                endLineNumber: next.line + 1,
+                startColumn: next.character,
+                startLineNumber: next.line + 1
+            };
+            const tRange = <theia.Range>Converter.toRange(range);
+            const searchResult: theia.TextSearchMatch = {
+                uri: URI.parse(result.fileUri),
+                preview: {
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    text: isObject(next.lineText) ? (<any>next.lineText).text : <string>next.lineText,
```
Comment (kittaakos) [inline packages/plugin-ext/src/plugin/workspace.ts:138]: Is this supposed to be a type-guard for `string | LinePreview`? If so, let's simplify to ` text: typeof next.lineText === 'string' ? next.lineText : next.lineText.text,` or use a proper type-guard and get rid of the `any`.

File: packages/plugin/src/theia-proposed.d.ts [hunk 10]
```diff
@@ -234,4 +234,133 @@ declare module '@theia/plugin' {
...
+        /**
+         * Whether or not to search for whole word matches only.
+         */
+        isWordMatch?: boolean;
+    }
+
+    /**
+     * Options that can be set on a findTextInFiles search.
+     */
+    export interface FindTextInFilesOptions {
```
Comment (kittaakos) [inline packages/plugin/src/theia-proposed.d.ts:272]: The VS Code API is different: https://github.com/microsoft/vscode/blob/8e1fe1d8999845ef8f88a6131a4a95b0f67fdd44/src/vs/vscode.proposed.d.ts#L681-L744

Which version are we targeting?

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts [hunk 11]
```diff
@@ -110,7 +114,11 @@ export class SearchInWorkspaceService implements SearchInWorkspaceClient {
         }
 
         const roots = await this.workspaceService.roots;
-        const searchId = await this.searchServer.search(what, roots.map(r => r.uri), opts);
+        return this.doSearch(what, roots.map(r => r.uri), callbacks, opts);
+    }
+
+    private async doSearch(what: string, rootsUris: string[], callbacks: SearchInWorkspaceCallbacks, opts?: SearchInWorkspaceOptions): Promise<number> {
```
Comment (kittaakos) [inline packages/search-in-workspace/src/browser/search-in-workspace-service.ts:120]: Should be `protected`.

**Comment by kittaakos** [comment]
I think I found the problem, the main issue here:
 - Search for something that does not hit the max count, it's important, your search won't be canceled,
 - Search for the same thing, you have duplicate results.

Please fix, and name me if you want a new review.

**Comment by nmorenor** [comment]
Just pushed the fixes for @kittaakos observations. Thanks for the review, I did not notice these issues, since I was just doing one search test.

File: packages/plugin/src/theia-proposed.d.ts [hunk 12]
```diff
@@ -234,4 +234,177 @@ declare module '@theia/plugin' {
...
+		/**
+		 * A [glob pattern](#GlobPattern) that defines the files to search for. The glob pattern
+		 * will be matched against the file paths of files relative to their workspace. Use a [relative pattern](#RelativePattern)
+		 * to restrict the search results to a [workspace folder](#WorkspaceFolder).
+		 */
+        include?: GlobPattern;
```
Comment (kittaakos) [inline packages/plugin/src/theia-proposed.d.ts:278]: Please check the formatting; you are mixing tabs with spaces.

File: packages/plugin-ext/src/common/plugin-api-rpc-model.ts [hunk 13]
```diff
@@ -539,7 +539,6 @@ export interface CallHierarchyOutgoingCall {
     to: CallHierarchyItem;
     fromRanges: Range[];
 }
-
```
Comment (kittaakos) [inline packages/plugin-ext/src/common/plugin-api-rpc-model.ts:542]: Let's revert this change.

**Comment by kittaakos** [comment]
I found a difference; search for `const index = `, see that leading whitespaces are ignored in Theia.

VS Code:
Screen Shot 2020-06-30 at 08 41 02


Theia:
Screen Shot 2020-06-30 at 08 41 15

**Comment by kittaakos** [comment]
👍 It works nicely; [one remark](https://github.com/eclipse-theia/theia/pull/7868#issuecomment-651579414) on the behavior and some minor formatting issues.

**Comment by nmorenor** [comment]
@kittaakos Formatting issues have been fixed also reverted the remove of line on plugin-api-rpc-model.ts

Now regarding the white spaces on the Theia, it looks like the search does return the line text with the white spaces there but somehow the UI is hiding the trailing white spaces. The extension does have a console.log for each returned result and in there I do see the trailing spaces but somehow they are not rendered on the UI. Perhaps that can be looked on a different PR? Here a Screen Shot of the console.log 

Screen Shot 2020-06-30 at 10 26 30 AM

I see this when I attach to the Host plugin while running on Debug.

File: packages/core/src/browser/style/sidepanel.css [hunk 14]
```diff
@@ -56,19 +56,19 @@
 }
 
 .p-TabBar.theia-app-left .p-TabBar-tab {
-    border-left: var(--theia-panel-border-width) solid transparent;
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/sidepanel.css:59]: @nmorenor I'm not sure which IDE you're using, but can you revert all the unnecessary changes like this?

Comment (nmorenor) [reply packages/core/src/browser/style/sidepanel.css]: Using vscode, I'm not sure why it changed the format on save, perhaps an extension. But those white space changes are already reverted.

**Comment by nmorenor** [comment]
Ok, now the tree also display the trailing whitespaces, I did add a new css class inside sidebar.css `noWrapInfoTree` I did not wanted to move the other one, because it could take effect on other parts of the UI.

**Comment by kittaakos** [comment]
> Ok, now the tree also display the trailing whitespaces, I did add a new css class inside sidebar.css `noWrapInfoTree` I did not wanted to move the other one, because it could take effect on other parts of the UI.

Wow, great 👍 I did not expect you to fix the tree too, but thank you! I'll take a look tomorrow, and will merge the changes.
